# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/xpconnect/loader/mozJSComponentLoader.cpp
# Commit: aca7cd850cc9
# Full Hash: aca7cd850cc9c3a88dfcd82d58268929aa288497
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2020-07-08 09:42:17
# Regressor Bug: 1627075
# File Overlap Count: 1
# Description:
#   Bug 1627075 - Route Omnijar requests through StartupCache r=froydnj
#   
#   This should be a relatively straightforward patch. Essentially, we implement
#   a wrapper class (and friends) around nsZipArchive (and friends), which transparently
#   caches entries from the underlying zip archive in the StartupCache. This will break
# ==============================================================================

diff -r d2e40366118c -r aca7cd850cc9 js/xpconnect/loader/mozJSComponentLoader.cpp
--- a/js/xpconnect/loader/mozJSComponentLoader.cpp	Wed Jul 08 02:46:11 2020 +0000
+++ b/js/xpconnect/loader/mozJSComponentLoader.cpp	Wed Jul 08 02:46:34 2020 +0000
@@ -51,6 +51,7 @@
 #include "mozilla/scache/StartupCacheUtils.h"
 #include "mozilla/MacroForEach.h"
 #include "mozilla/Preferences.h"
+#include "mozilla/Omnijar.h"
 #include "mozilla/ResultExtensions.h"
 #include "mozilla/ScriptPreloader.h"
 #include "mozilla/ScopeExit.h"
@@ -670,6 +671,10 @@
 
 static mozilla::Result<nsCString, nsresult> ReadScript(
     ComponentLoaderInfo& aInfo) {
+  // We're going to cache the XDR encoded script data - suspend writes via the
+  // CacheAwareZipReader, otherwise we'll end up redundantly caching scripts.
+  AutoSuspendStartupCacheWrites suspendScache;
+
   MOZ_TRY(aInfo.EnsureScriptChannel());
 
   nsCOMPtr<nsIInputStream> scriptStream;